cpp/serve/engine_actions/batch_verify.cc

```diff
@@ -0,0 +1,403 @@
+
+    // - Get embedding and run verify.
+    Array<NDArray> embeddings;
+    std::vector<int64_t> request_internal_ids;
+    std::vector<int> verify_lengths;
+    int total_verify_length = total_draft_length;
+    embeddings.reserve(num_requests);
+    request_internal_ids.reserve(num_requests);
+    verify_lengths.resize(num_requests);
+    for (int i = 0; i < num_requests; ++i) {
+      RequestModelState verify_mstate = rstates[i]->mstates[verify_model_id_];
+      RequestModelState draft_mstate = rstates[i]->mstates[draft_model_id_];
+      ICHECK_EQ(draft_mstate->GetDraftLength(), draft_lengths[i]);
+      // one committed token and all the draft tokens
+      verify_lengths[i] = draft_lengths[i];
+      ICHECK(draft_mstate->draft_output_tokens.size() ==
+             draft_mstate->draft_output_token_prob.size());
+      ICHECK(draft_mstate->draft_output_tokens.size() ==
+             draft_mstate->draft_output_prob_dist.size());
+      request_internal_ids.push_back(verify_mstate->internal_id);
+      RECORD_EVENT(trace_recorder_, requests[i]->id, "verify start embedding");
+      std::vector<int> draft_tokens = {draft_mstate->committed_tokens.back()};
+      draft_tokens.insert(draft_tokens.end(), draft_mstate->draft_output_tokens.begin(),
+                          draft_mstate->draft_output_tokens.end());
+      if (draft_tokens.size() > 1U) {
+        draft_tokens.pop_back();
+      } else {
+        verify_lengths[i] += 1;
+        total_verify_length += 1;
+      }
+      embeddings.push_back(models_[verify_model_id_]->TokenEmbed(
+          {IntTuple{draft_tokens.begin(), draft_tokens.end()}}));
+      RECORD_EVENT(trace_recorder_, requests[i]->id, "verify finish embedding");
+    }
+
+    RECORD_EVENT(trace_recorder_, request_ids, "start verify");
+    NDArray logits =
+        models_[verify_model_id_]->BatchVerify(embeddings, request_internal_ids, verify_lengths);
+    RECORD_EVENT(trace_recorder_, request_ids, "finish verify");
+    ICHECK_EQ(logits->ndim, 3);
+    ICHECK_EQ(logits->shape[0], 1);
+    ICHECK_EQ(logits->shape[1], total_verify_length);
+
+    std::vector<int> cum_verify_lengths = {0};
+    for (int i = 0; i < num_requests; ++i) {
+      cum_verify_lengths.push_back(cum_verify_lengths.back() + verify_lengths[i]);
+    }
+
+    // - Verify tokens.
+    logits = logits.CreateView({total_verify_length, 1, logits->shape[2]}, logits->dtype);
+    std::vector<RequestModelState> mstates;
+    std::vector<GenerationConfig> generation_configs;
+    int cur_req_id = 0;
+    for (int i = 0; i < total_verify_length; ++i) {
+      if (i >= cum_verify_lengths[cur_req_id + 1]) {
+        cur_req_id += 1;
+      }
+      Request request = requests[cur_req_id];
+      mstates.push_back(estate->GetRequestState(request)->mstates[verify_model_id_]);
+      generation_configs.push_back(request->generation_cfg);
+    }
+
+    logits = sampler_->ComputeProb(logits, models_[verify_model_id_], mstates, generation_configs);
```

Comment:
I am a bit hesitating about whether we can compute prob for all positions in all requests altogether here. The main concern is when frequency/presence penalty need to apply, we do not have the precise logits for position `i + 1` before sampling the token at position `i`. It would be okay to do in the current way without frequency/presence penalty.

Reply:
I have rewritten this part to compute probs one by one. But there are redundant memory copies between device and host. Is there any way to view a slice of subarray of a NDArray directly on device?